### ISSUE TITLE
How to avoid uncatchable error handling

### ISSUE THREAD

**Issue by mikeyaworski** [microlinkhq/tinyspawn#13]
I'm a bit confused by the code here:

https://github.com/microlinkhq/tinyspawn/blob/7659647f01ecef03e6033ae30c81a3022fec46aa/src/index.js#L40-L62

You are attaching a listener to the child process which force throws an error. Maybe I'm missing something, but I have no way to catch this error, so my server just force crashes as a result.

My current workaround is to just call

```
const process = ...
process.removeAllListeners('error');
process.removeAllListeners('exit');
process.on('error', () => {
  // my own error handling
});
```

The code I've just written is not good, since I don't technically know what I'm removing, which can lead to a memory leak. Is there an alternative I'm missing? I would like to not crash my server whenever there is an error.

Along the same lines, how do I force the child process to exit gracefully (there is no error - I just want it to terminate)?
- `process.kill()` will cause a thrown error that I mentioned above
- `process.emit('exit', 0, null)` is what I figure is the correct answer

TL;DR:
1. How do I prevent the process crashing my server with a thrown error whenever an error event happens?
2. How do I gracefully terminate the process?

**Comment by Kikobeats** [comment]
Thanks for reporting; Can you tell me if https://github.com/microlinkhq/tinyspawn/pull/14 reproduces the behavior you are describing?

I will need to modify the codebase for wrapping `process.kill` in a better way that keeps consistency with the exposed promise interface.

**Comment by mikeyaworski** [comment]
Yes that reproduces the behavior.

Just to be clear, I also want to prevent an (uncatchable) error being thrown whenever the process itself has an error, not just when calling `process.kill`.

**Comment by Kikobeats** [comment]
errors are being captured, see this test https://github.com/microlinkhq/tinyspawn/blob/c1ebcb361ebe355693dbff88bc84832855afe87b/test/index.js#L15-L23

I believe the only thing buggy is to kill the process explicitly while running the promise at the same time.

**Comment by Kikobeats** [comment]
@mikeyaworski you are expecting unhandled rejections because the promise is floating until you await for it
https://github.com/microlinkhq/tinyspawn/pull/14/files#diff-5bb8db779819ddef5956a5d9d5949c05ef7445237656ca37bf2f02720271440bR31

**Comment by Kikobeats** [comment]
Also you can pass timeout and kill signal to the process, I think that what you want at the end:
https://github.com/microlinkhq/tinyspawn/blob/7659647f01ecef03e6033ae30c81a3022fec46aa/test/index.js#L128-L136

**Comment by mikeyaworski** [comment]
Ah okay. I think the problem is actually in the `youtube-dl-exec` library that you have, so I will create an issue over there instead. I didn't notice that this library returned a promise. It is the other library where I want to handle the unhandled rejection.
